### PR TITLE
Add condition to VirtualMachineVolume status

### PR DESCRIPTION
**Pull request by hbinkim**
Closes: #1

This pull request adds condition array to VirtualMachineVolume status following [the recent guideline ](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1623-standardize-conditions) proposed in k8s sig-api-machinery group to standardize condition usage. 

Due to VirtualMachineVolume CRD change, general code refactoring had to take in place.

### CODE REVIEW

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolume_types.go [hunk 1]
```diff
@@ -33,18 +33,43 @@ const (
...
+	Conditions []VirtualMachineVolumeCondition `json:"conditions,omitempty"`
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // VirtualMachineVolume is the Schema for the virtualmachinevolumes API
 // +kubebuilder:subresource:status
 // +kubebuilder:resource:path=virtualmachinevolumes,scope=Namespaced,shortName=vmv
-// +kubebuilder:printcolumn:name="State",type="string",JSONPath=".status.state",description="Current state of VirtualMachineVolume"
+// +kubebuilder:printcolumn:name="Status",type="string",JSONPath=".status.conditions[?(@.type==\"ReadyToUse\")].status",description="Current status of VirtualMachineVolume"
```
Comment (woohhan) [inline pkg/apis/hypercloud/v1alpha1/virtualmachinevolume_types.go:71]: state gives an overview of the conditions. The State is better for showing to the user

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolume_types.go [hunk 2]
```diff
@@ -33,18 +33,43 @@ const (
...
+// VirtualMachineVolumeConditionType is a string representation of known VirtualMachineVolume conditions
+type VirtualMachineVolumeConditionType string
+
+const (
+	// ReadyToUse indicates VirtualMachineVolume is ready to use
+	ReadyToUse VirtualMachineVolumeConditionType = "ReadyToUse"
+)
+
+// VirtualMachineVolumeCondition contains details of the current condition of a VirtualMachineVolume.
+type VirtualMachineVolumeCondition struct {
```
Comment (woohhan) [inline pkg/apis/hypercloud/v1alpha1/virtualmachinevolume_types.go:45]: looks better to provide a generic structure for the condition.

File: e2e/virtualmachinevolume.go [hunk 3]
```diff
@@ -3,6 +3,7 @@ package e2e
 import (
 	"context"
 	goerrors "errors"
+	"github.com/google/go-cmp/cmp"
```
Comment (woohhan) [inline e2e/virtualmachinevolume.go:6]: Why this package is needed? looks reflect.deepEqual is better

File: e2e/virtualmachinevolume.go [hunk 4]
```diff
@@ -12,7 +13,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/wait"
 	"kubevirt-image-service/pkg/apis"
 	hc "kubevirt-image-service/pkg/apis/hypercloud/v1alpha1"
-	volume "kubevirt-image-service/pkg/controller/virtualmachinevolume"
+	v "kubevirt-image-service/pkg/controller/virtualmachinevolume"
```
Comment (woohhan) [inline e2e/virtualmachinevolume.go:16]: The package name is used for readability. see this link https://rakyll.org/style-packages/

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolume_types.go (hunk 2 above)
Comment (woohhan) [reply pkg/apis/hypercloud/v1alpha1/virtualmachinevolume_types.go:45]: #15 I will handle this issue here.

File: pkg/apis/hypercloud/v1alpha1/virtualmachinevolume_types.go (hunk 1 above)
Comment (hbinkim) [reply pkg/apis/hypercloud/v1alpha1/virtualmachinevolume_types.go:71]: done

File: e2e/virtualmachinevolume.go (hunk 3 above)
Comment (hbinkim) [reply e2e/virtualmachinevolume.go:6]: done

**Comment by hbinkim** [comment]
Applied condition array util. Need another code review for this change. @woohhan

**Comment by woohhan** [comment]
please review the commit @hyoung-90

File: pkg/controller/virtualmachinevolume/virtualmachinevolume_controller_test.go [hunk 5]
```diff
@@ -54,22 +42,15 @@ var _ = Describe("Reconcile", func() {
...
+			Expect(found).Should(BeTrue())
+			Expect(cond.Status).Should(Equal(corev1.ConditionFalse))
 		})
 	})
 
 	Context("2. with false status image", func() {
 		image := newTestImage()
-		readyToUse := false
-		image.Status.ReadyToUse = &readyToUse
+		util.SetConditionByType(image.Status.Conditions, hc.ConditionReadyToUse, corev1.ConditionTrue, "VmiIsReady", "Vmi is ready to use")
```
Comment (hyoung-90) [inline pkg/controller/virtualmachinevolume/virtualmachinevolume_controller_test.go:53]: It is a test case where the status of the image is false, but the value put in image condition seems to be true.

Comment (hbinkim) [reply pkg/controller/virtualmachinevolume/virtualmachinevolume_controller_test.go:53]: Thank you for spotting the error! It is fixed now.

**Comment by hbinkim** [comment]
@woohhan conflict between master branch is resolved

**Pull request merged**